message_feed: Enhance subscription management with Subscribe link and streamline status display.

Sreejan09 opened the pull request:
This commit adds a Subscribe option, allowing users to join a channel with a single click. It also removes redundant subscription status information, improving the overall user experience and interface clarity.

Fixes: #33060

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

 - [x] Add a "Subscribe" link at the end of the "You are not subscribed to {channel}." divider, that subscribes the user to the current channel.
 - [x] Only show links on the last channel subscription status divider, not on all of them.

## Screenshots for the Subscribe Link Option:


|  Before  |  After  |
|  ---  |  ---  |
| ![before_subscribeLink](https://github.com/user-attachments/assets/9d24b2c4-9832-4206-a2dd-d56256222c31) | ![After Subscribe Link](https://github.com/user-attachments/assets/9c7ae697-7d7e-4c96-855e-644050d5aa75) |

## Screenshots for Redundant Subscription Status:

|  Before  |  After  |
|  ---  |  ---  |
| ![Before](https://github.com/user-attachments/assets/52114e67-01d3-41b2-b150-f1ec29a061c7) | ![After](https://github.com/user-attachments/assets/80f73e70-2de4-4d2c-8d38-05f177ec6845) |
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Demo video for the Subscribe Link Option</summary>

#### Before Subscribe Option

 https://github.com/user-attachments/assets/d525b906-146d-4da7-aa40-34a8a44c14fe  


#### After Subscribe Option

https://github.com/user-attachments/assets/80ea4d35-0a4e-480f-ab16-8ee32a007847 

</details>

<details>
<summary>Demo video for the Redundant Subscription Status</summary>


#### Before: 

https://github.com/user-attachments/assets/790a4c9c-ebde-4214-bf0b-2a7eca675a7e

#### After: 

https://github.com/user-attachments/assets/ae0fff6c-d61f-41e8-b5e6-448e93dd695a

</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Please include a screenshot of the situation described in the issue. These screenshots don't currently demonstrate:

> Add a "Subscribe" link at the end of the "You are not subscribed to {channel}." divider, that subscribes the user to the current channel.

You'll also need to complete the self-review checklist in order to prepare this PR for review.

Sreejan09 commented:
Hey @alya , could you review my PR? Thanks!

Sreejan09 commented:
Hey @alya , could you review my PR? Thanks!

alya commented:
@laurynmm Could you please take a look at this one? Not release-critical, but would be nice to fix if it's easy to do so.

laurynmm commented:
@Sreejan09 - Thanks for working on this issue!

These changes break the current bookend shown for the spectator view (aka when someone who's not logged in views a [public access channel](https://zulip.com/help/public-access-option)). See before/after screenshots below. You'll want to carefully test your changes manually to make sure you're not introducing any regressions before pinging me for another review.

Also, you should review our documentation for [writing commit messages](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages). You current commit message isn't formatted correctly and the description could be improved.

---

| Spectator view of channel | Screenshot |
| --- | --- |
| Before | ![Screenshot from 2025-02-20 14-12-41](https://github.com/user-attachments/assets/af5fe660-bfc7-4aa7-90cf-1ffb3e5333a1) |
|After |  ![Screenshot from 2025-02-20 14-10-25](https://github.com/user-attachments/assets/63aa7f0e-a9f7-47f9-916e-dcb054630430) |

Sreejan09 commented:
Hi @laurynmm,
I've updated the commit message to align with the guidelines and fixed the spectator view issue. I also manually tested the changes to ensure everything worked as expected. Could you please review it again?

Thanks!

laurynmm commented:
@Sreejan09 - So, #33060 does not say to remove the channel subscription dividers unless they are the final, most recent one (aka the trailing bookend). It says to "only show links on the last channel subscription status divider, not on all of them".

Your changes currently change how these dividers are rendered so that only the last one is rendered at all. So the changes shown in your "redundant subscription status" screenshots above are incorrect. The correct change would be to remove the linked text from all but the final divider.

amanagr commented:
Thanks for your work @Sreejan09! Closing due to too many unrequested changes in the PR. I would have helped but we are close the release and hence opened #33646 to fix the issue.
